### ISSUE TITLE
[hipri] Gantt improvements - improve default view ranges, add interactivity

### ISSUE THREAD

**Issue by cleary** [DeBortoliWines/dotProject#7]
Hi guys, 
Opening this one up for design/discussion - 
Our PD (Product Development) team has highlighted this as a high value requirement for their planning, and hopefully re-scheduling capabilities. 

Currently the Gantt chart included in dP is non-interactive, and doesn't make sensible choices about default date ranges for viewing a project, so it frequently requires checking dates of tasks etc.

The major features they're looking for are: 
Stage 1:
* Better visualisation control and interactivity

Stage 2:
* Project editing access - rescheduling tasks and dependencies

Stage X wishlist items:
* export to ... options

I had a bit of a look around for open source javascript Gantt Chart libraries, and discovered [frappe gantt](https://frappe.io/gantt) - it's open source, and looks to solve most or all of the above requirements out of the box.

Can I get some discussion/recommendations on how best to implement this initially @ajdonnison - would you prefer to see it as a module while we're still getting it tried and tested before considering replacement of the existing core gantt code?

@mattbell87 can you take a look at this one please?

**Comment by cleary** [comment]
@rafaelqg - it's potentially worth you looking at this one too and how you might affect/compliment your current branch/development focus

**Comment by mattbell87** [comment]
Frappe looks nice, I wonder if it's themeable. I'm interested to see what @ajdonnison thinks. It looks like it runs off of events, so we may need some sort of php API that listens to xhr calls in order to save to the DB.

**Comment by rafaelqg** [comment]
@cleary Thanks for let me know about this new branch.
I don't know if @mattbell87  is going to take this assignment, but if necessary I can handle this as well

**Comment by rafaelqg** [comment]
I didnt knew frappe (https://frappe.io/gantt), it looks very suitable for our demands.

**Comment by cleary** [comment]
Thanks for the replies guys, 

@rafaelqg thanks for the offer, I'll get @mattbell87 to lay the foundational elements for now while you and I finalise the other projects you're working on. Once they're completed we can revisit

@mattbell87 I just spoke to our PD manager again, I've proposed that a non-interactive visualisation module may be a relatively simple thing to start with (ie, just display the project in Gantt form, and disable any task change features). 
What are your thoughts on creating a frappe gantt view-only module as a starting point? 
Are there any other gantt library alternatives that we should be considering?
Do we need to dig a little deeper into frappe gantt's features just to confirm if it's going to get us to our endpoint?

@ajdonnison do you have any thoughts on this proposal?

**Comment by ajdonnison** [comment]
A quick look suggests that it would be a great resource. It should be relatively easy to set up a fully interactive version. You'd just need to tie the event listeners to XHR calls back to adjust the source item values. Security would need to be thought through on the calling mechanism, but apart from that I see it as something well worth including.

Adam Donnison
ajdonnison@gmail.com (https://link.getmailspring.com/link/2AE40D08-5FD6-4A14-BF13-EC18C2738B8C@getmailspring.com/0?redirect=mailto%3Aajdonnison%40gmail.com&recipient=cmVwbHkrMDAwMTQ5OGI4MjQ4Y2IxOGMwN2U1ZDEzZmZlNTlmY2FlMGEyMzAxZTMxZWZiYmY4OTJjZjAwMDAwMDAxMTg5OTZjYmU5MmExNjljZTE4ZTMwZTNjQHJlcGx5LmdpdGh1Yi5jb20%3D)
REDBANK VIC 3477 Australia (https://link.getmailspring.com/link/2AE40D08-5FD6-4A14-BF13-EC18C2738B8C@getmailspring.com/1?redirect=https%3A%2F%2Fmaps.google.com%2F%3Fq%3DREDBANK%2520VIC%25203477%2520Australia&recipient=cmVwbHkrMDAwMTQ5OGI4MjQ4Y2IxOGMwN2U1ZDEzZmZlNTlmY2FlMGEyMzAxZTMxZWZiYmY4OTJjZjAwMDAwMDAxMTg5OTZjYmU5MmExNjljZTE4ZTMwZTNjQHJlcGx5LmdpdGh1Yi5jb20%3D)

**Comment by cleary** [comment]
Thanks @ajdonnison :)

@mattbell87 can I get you to start on this as a matter of priority please? 
The first stage goal I'd like to aim for is just a read-only visualisation of a project in frappe gantt as a standalone module

**Comment by cleary** [comment]
Hi @mattbell87 - just confirming you're working on this? I can see there's a branch (but currently no commits) - https://github.com/DeBortoliWines/dotProject/tree/7-gantt-improvements

**Comment by mattbell87** [comment]
Yep I have some stuff that's not pushed

**Comment by cleary** [comment]
Fantastic, thanks :)

**Comment by mattbell87** [comment]
Hi @cleary I currently have Frappe gantt rendering on the project list page. I will begin to work on the gantts for tasks within a project.
<img width="984" alt="Screen Shot 2019-03-20 at 8 44 26 pm" src="https://user-images.githubusercontent.com/2722612/54674699-18adfa80-4b51-11e9-82d9-3d31fcd5db54.png">
Currently there are controls from the old system above frappe but they don't work. I'm guessing these controls are particularly useful when you have a lot of projects and you want to filter them down.

**Comment by cleary** [comment]
Thanks Matt, really appreciate the progress update!

I've done a test build here, and I'm getting the following error:
```
Parse error: syntax error, unexpected 'const' (T_CONST), expecting variable (T_VARIABLE) in /srv/dotProject/modules/projects/frappegantt.php on line 12
```
This is CentOS 7.4 with PHP5.4  - I know it's not ready to deploy, but just a heads up (in case there's some specific php version exclusion or similar that's happening)

![Screenshot_20190321_093927](https://user-images.githubusercontent.com/2577966/54723935-3a44cb80-4bbd-11e9-96ee-7fa9b37f42b9.png)

**Comment by mattbell87** [comment]
Ah looks like the visibility modifier (public) on the class const only works in PHP > 7.1.0 according to php.net. I'll fix that up tonight. Thanks.

**Comment by mattbell87** [comment]
Hey again @cleary please have a look at the latest changes in [7-gantt-improvements](https://github.com/DeBortoliWines/dotProject/tree/7-gantt-improvements). Projects and tasks are working now and you can click items within gantts and see the progress of tasks. I have removed some of the checkboxes that were more specific to the old charting library. Also since there's no built in method I've made it read only by resetting the chart if anything has been changed.

**Comment by cleary** [comment]
Hi Matt, 
Apologies this has taken me a bit of time to get focus on - 

Thanks for the update, I've had a bit of a play and it's looking pretty close to functional. 
I'm not sure if you're after bug reports yet, but I've got a couple of notes:

1. It's not loading from the Gantt tab in a Project view until I click "Show Full Project"
![Screenshot_20190409_163059](https://user-images.githubusercontent.com/2577966/55777967-df7c0100-5ae4-11e9-9868-844418a67f9e.png)

2. The date adjustment handles on tasks are active but don't do anything (I assume you've left this in while it's a WIP though)

3. The default date ranges chosen for "Show full project" are not useful, I don't know if you want to leave this as a separate thread, but it was one of our primary bugbears with the previous gantt module. It mostly just shows white screen and I have had to go digging/scrolling/sliding to find my tasks

4. The fixed zoom is problematic - all the "real life" projects I looked at had tasks which ran significantly longer than the default page width would handle. 

5. Adding to 4., for each bar on the chart, the task title is centred, which means for very long tasks I have to scroll a lot just to see what the task is 

![Screenshot_20190409_164652](https://user-images.githubusercontent.com/2577966/55778749-15ba8000-5ae7-11e9-8e85-c8d6d049d33c.png)

6. It's showing tasks from unrelated projects (which probably exacerbates the problem in 3.)

**Comment by mattbell87** [comment]
Hey @cleary,

Bug reports are always welcome! I'll have a look at each issue and see what I can do to fix it. 

The month handles were working for me last time I checked, not sure what's going on there. I'd check again now but it looks like I have to fix my mariadb server again (stupid macs!). 

It looks like frappe has some nice zoom levels built in, so I'll look at enabling them.

**Comment by cleary** [comment]
Thanks @mattbell87 :)

**Comment by mattbell87** [comment]
Hey @cleary as above I've improved a lot of the code to try and cover the points you made earlier. I think from what you said the primary concern was too much scrolling. So I've added in the view functionality of Frappe by adding some buttons. I've also made each gantt chart individually remember what view mode it was on last, this should help with some projects being long term and others being short term. 

I'm fairly sure I've also corrected some of the weirdness with tasks from other projects being shown and other tasks being missing. I don't have a lot of data to test with so you may find otherwise.

**Comment by cleary** [comment]
Thanks @mattbell87 - apologies I've been off the grid the last week with a sick family and got some other hipri projects on the boil. 
I'll give this a good test this week though (hopefully!)

**Comment by mattbell87** [comment]
Got it no problems :)

**Comment by cleary** [comment]
Hi @mattbell87 - 
Once again, apologies for the late reply -
I found a few issues, and was having difficulty articulating what I was seeing -

Some general impressions:
 - the zoom controls (quarter day/half day/week...year) are a great addition, that will be really handy
 - the re-schedule handles are active (edges to change start/end date, or move task) and look really promising (I understand they're not functional yet)
 - the project gantt view (tab accessed from the projects page) also looks very promising as an overview feature

Where I'm still having issues:
* our existing projects just don't seem to show any tasks - if I create a new project and some new tasks that are relevant now, they show up ok, but it's the really big running projects that I want to look at
* frappe is still not sensibly placing titles in the visible frame. It just centers the title on the bar, regardless of visibility (and frequently the tasks are long and the centrepoint is off screen)

I think at this point I will need to share a database dump with you, so that you can test for yourself with real projects - is that ok?

**Comment by mattbell87** [comment]
Hi @cleary,

Yeah absolutely a database dump would help a lot to put things in to perspective!

**Comment by cleary** [comment]
Hey @mattbell87 - thanks for the offer, given the sensitive nature of some of the projects I've given @Givo29 (Joel) the chance to have a look at some of the problems. Joel is at uni and works 2 days a week for us in the office so he has direct access already -

If he runs into any dramas we'll follow up with you again - 

As a heads up, the commits that went in today were a rebase from our (not upstream) master branch which includes a few permission fixes from upstream/dev (as well as a patch for the visibility issues).

**Comment by mattbell87** [comment]
No problems I understand. Perhaps in the future we should talk about putting together a large amount of dummy data that everybody can use, that would be useful for anyone developing this application.

**Comment by cleary** [comment]
That sounds like a good plan... 
@rafaelqg @ajdonnison are you aware of/do you have anything like this?

**Comment by mattbell87** [comment]
Hey @cleary this one was closed however there was no PR opened for it afaik, at least in to the [devel](/dotproject/dotProject/tree/devel) branch.

Also what are your current thoughts on making it interactive?

**Comment by cleary** [comment]
Hey Matt,
@Givo29 got it into shape for our minimum requirements, it's been deployed internally and we're doing a lot of testing. There was some weird behaviour here and there that we needed to resolve before a PR was appropriate - it's in pretty good shape as of this afternoon though...

I know Joel is interested in pursuing the interactivity side as well, but if you've got time and interest tonight to get stuck in, please do. 

Now that we're knocking over some of the big hurdles I really need to revisit my list... (this is not a complaint btw 😅)

**Comment by mattbell87** [comment]
Awesome. I think I'll let @Givo29 clean up things before starting more work. It looks like theres more to be merged. Maybe merging the current state of Frappe in to dotproject/devel would be a good place to start before looking at the interactive version. 

When we do build the interactive version we'll need to think about a JSON api for the JS to interact with and also think about keeping it all solid and secure. When it's time we might discuss that with @ajdonnison in a new issue.
